cmd/publicize/server.go

```diff
@@ -0,0 +1,188 @@
+
+type githubClient interface {
+	IsMember(org, user string) (bool, error)
+	CreateComment(owner, repo string, number int, comment string) error
+	GetPullRequest(org, repo string, number int) (*github.PullRequest, error)
+
+	ListIssueComments(owner, repo string, number int) ([]github.IssueComment, error)
+}
+
+var publicizeRe = regexp.MustCompile(`(?mi)^/publicize\s*$`)
+
+func helpProvider(_ []prowconfig.OrgRepo) (*pluginhelp.PluginHelp, error) {
+	pluginHelp := &pluginhelp.PluginHelp{
+		Description: `The publicize plugin is used for merging a pull request into the bonded repository.`,
+	}
+	pluginHelp.AddCommand(pluginhelp.Command{
+		Usage:       "/publicize",
+		Description: "Merge the specific pull request into the bonded repository",
+		WhoCanUse:   "Members of the trusted organization for the repo.",
+		Examples:    []string{"/publicize"},
+	})
+	return pluginHelp, nil
+}
+
+type server struct {
+	githubTokenGenerator func() []byte
+
+	gitName  string
+	gitEmail string
+
+	bonds map[string]string
+
+	ghc githubClient
+	gc  git.ClientFactory
+
+	remoteHost string
+
+	dry bool
+}
+
+func (s *server) handleIssueComment(l *logrus.Entry, ic github.IssueCommentEvent) {
+	if !publicizeRe.MatchString(ic.Comment.Body) {
+		return
+	}
+
+	org := ic.Repo.Owner.Login
+	repo := ic.Repo.Name
+	num := ic.Issue.Number
+	commentAuthor := ic.Comment.User.Login
+
+	logger := logrus.WithFields(logrus.Fields{
+		github.OrgLogField:  org,
+		github.RepoLogField: repo,
+		github.PrLogField:   num,
+	})
+
+	if !ic.Issue.IsPullRequest() {
+		s.createComment(ic, fmt.Sprintf("Publicize plugin can only be used in pull requests"), logger)
+		return
+	}
+
+	// Only org members should be able to publicize a pull request.
+	ok, err := s.ghc.IsMember(org, commentAuthor)
+	if err != nil {
+		return
+	}
+	if !ok {
+		s.createComment(ic, fmt.Sprintf("only [%s](https://github.com/orgs/%s/people) org members may request a publicize", org, org), logger)
+		return
+	}
+	logger.Info("Publicize of PR has been requested.")
+
+	pr, err := s.ghc.GetPullRequest(org, repo, num)
+	if err != nil {
+		logger.WithError(err).Warn("couldn't get pull request")
+		return
+	}
+	baseBranch := pr.Base.Ref
+
+	if !pr.Merged {
+		s.createComment(ic, "cannot publicize an unmerged pull request", logger)
+		return
+	}
+
+	logger.Info("Searching bond...")
+	destOrgRepo, ok := s.bonds[fmt.Sprintf("%s/%s", org, repo)]
+	if !ok {
+		logger.Warn("There is no public repository configured for the current repository.")
+		s.createComment(ic, "cannot publicize because there is no bond for the current pull request", logger)
+		return
+	}
+
+	destOrg := strings.Split(destOrgRepo, "/")[0]
+	destRepo := strings.Split(destOrgRepo, "/")[1]
+	sourceRemoteResolver := func() (string, error) {
+		u, err := url.Parse(s.remoteHost)
+		if err != nil {
+			return "", err
+		}
+
+		// If the host isn't a url but its an actual path, return the full path with org and repo included.
+		// This is usefull for testing as well.
```

Comment:
This smells. In what case does this branch run during production execution?

Reply:
This is a little bit hacky. I will open a follow-up upstream PR to make the `FetchFromRemote` more testable.  Currently, we need to pass a `remoteResolver` and to test this we need a resolver from an actual path.

Reply:
Is this resolved? I don't want us to have special code in here that's only used in manual testing

Reply:
This will rely on the git v2 design here. I will do it as a follow up since some investigation is needed.